### PR TITLE
Add support for Plex Watchlist importing

### LINKED ISSUE #5705: Feature Request: Plex Watchlist monitoring
Plex has been rolling out and improving upon a personal watchlist feature. It used to only work with their own stream-able content. But it now works with any content at all. You can manage your own watchlist at https://mediaverse.plex.tv/

I'd like to add an additional `List` Subheader and list item for `Plex` in `Settings > Lists`. The list API needs an authentication token. So I'd either re-use the `Settings > Connect > Plex` token or provide an additional login feature on the list item settings panel.

Here are some rough ideas on what the screens would look like:

modal
lists

[AB#664](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/664)

### LINKED ISSUE #7187: Pull Sonarr commit 'Fixed: Error when trying to import an empty Plex Watchlist'
Pull https://github.com/Sonarr/Sonarr/commit/6e271e9272d6a109bc4b8f3debe784d767ee7a63

```
git cherry-pick -ex 6e271e9272d6a109bc4b8f3debe784d767ee7a63
```

This commit has 0 conflict(s)

You can ask me to `close all`, `open pr here` or `open pr everywhere`

[AB#2913](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/2913)

### PR DESCRIPTION
**Pull request by vanstinator**
#### Database Migration
NO

#### Description

Plex has a new watchlist feature that they've been fleshing out. I've been using it and wanted to see support in Radarr. I've implemented a basic import list implementation taking cues from the existing trakt integration. Where possible I've re-used or enhanced existing Plex response models rather than write new code. I'm a node dev professionally and haven't written C# in years so any and all feedback is very welcome.

#### Todos
- [ ] Tests
- [ ] Translation Keys
- [ ] Wiki Updates

#### Issues Fixed or Closed by this PR

* Closes #5705 
* Closes #7188 
* Closes #7187

### CODE REVIEW

File: src/NzbDrone.Core/ImportLists/Plex/List/PlexListSettings.cs
Comment (vanstinator) [inline src/NzbDrone.Core/ImportLists/Plex/List/PlexListSettings.cs:1]: I'm not actually sure I need this file at all as the base settings contain... everything

File: src/NzbDrone.Core/ImportLists/Plex/PlexImportBase.cs
Comment (vanstinator) [inline src/NzbDrone.Core/ImportLists/Plex/PlexImportBase.cs:47]: The function body here is a direct copy of one of the other plex handlers. It probably makes sense to shove this in a module somewhere so we have one source of truth.

**Comment by bakerboy448** [comment]
I’ll save other comments for the devs, but you have translating todo

-apply translate function / create Keys
-add English Keys and their text to en.json

File: src/NzbDrone.Core/ImportLists/Plex/PlexSettingsBase.cs
Comment (vanstinator) [inline src/NzbDrone.Core/ImportLists/Plex/PlexSettingsBase.cs:50]: Out of scope for this PR. But it might make sense to make a common import list settings base for all these filters that seem to be somewhat common in all lists.

File: src/NzbDrone.Core/Notifications/Plex/PlexTv/PlexTvService.cs
Comment (vanstinator) [inline src/NzbDrone.Core/Notifications/Plex/PlexTv/PlexTvService.cs:85]: There really wasn't a great place to put this without writing a bunch of boilerplate code to wrap plex media providers.

File: src/NzbDrone.Core/Notifications/Plex/Server/PlexSectionItem.cs
Comment (vanstinator) [inline src/NzbDrone.Core/Notifications/Plex/Server/PlexSectionItem.cs:9]: `ratingKey` is considered a string in the Plex API. And the values coming back from providers are a uuid rather than an int. So this change is required to make the list import work.

**Comment by vanstinator** [comment]
@bakerboy448 I'm looking at the other import lists and none of them are using the translation service. I'm only seeing translation keys for the frontend code and I didn't write any frontend code here.

**Comment by austinwbest** [comment]
> @bakerboy448 I'm looking at the other import lists and none of them are using the translation service. I'm only seeing translation keys for the frontend code and I didn't write any frontend code here.

There are translations done for the back end as well, just not in the same method as the front end. Not everything has been translated yet, but wouldn't hurt to apply it to new code

**Comment by vanstinator** [comment]
> > @bakerboy448 I'm looking at the other import lists and none of them are using the translation service. I'm only seeing translation keys for the frontend code and I didn't write any frontend code here.
> 
> There are translations done for the back end as well, just not in the same method as the front end. Not everything has been translated yet, but wouldn't hurt to apply it to new code

Perfect. I'll take a look and get the Plex code translated. If it's simple enough I can translate some of the other code I used as inspiration after this merges.

**Comment by vanstinator** [comment]
I'm not familiar enough with .NET or TinyIOC to refactor the settings classes in such a way as to allow passing the localization service as a constructor arg. I'm happy to help. But it's starting to get out of scope for this specific PR.

I've also deleted the majority of the settings that were specific to Trakt. So the scope of what actually needs translating is quite low at this point.

**Comment by vanstinator** [comment]
Sorry I wasn't clear in the earlier post. I cannot localize this as I'm not familiar enough with the tools. I tried injecting the localization service but received compiler errors when attempting to use it in the `FieldDefinition` label. I need help to get to a point where it works. And once I understand what needs to happen I would be more than happy to offer my assistance in localizing the current import lists as a thank you.

**Review by Qstick — CHANGES_REQUESTED** [review]
Looks good overall, quick scan found a few issues tagged. Also, why the perms change on CustomScript.cs?

`
src/NzbDrone.Core/Notifications/CustomScript/CustomScript.cs  100755 → 100644
`

Don't worry about the translations for now, we havn't done any for backend provider settings and I'd rather do them separately.

File: src/NzbDrone.Core.Test/ImportListTests/Plex/PlexParserFixture.cs
Comment (vanstinator) [reply src/NzbDrone.Core.Test/ImportListTests/Plex/PlexParserFixture.cs:10]: 2 guesses on which provider I copied 😛

File: src/NzbDrone.Core/ImportLists/Plex/PlexParser.cs
Comment (vanstinator) [reply src/NzbDrone.Core/ImportLists/Plex/PlexParser.cs:41]: Not yet. It's a bug on their end for this endpoint. The fix is shipping next week. Didn't want to hold up on getting feedback on this body of work. Once I've confirmed the fix actually ships I'll issue another PR adding any relevant IDs.

**Comment by vanstinator** [comment]
> Looks good overall, quick scan found a few issues tagged. Also, why the perms change on CustomScript.cs?
> 
> `src/NzbDrone.Core/Notifications/CustomScript/CustomScript.cs 100755 → 100644`
> 
> Don't worry about the translations for now, we havn't done any for backend provider settings and I'd rather do them separately.

Arg. I use WSL 2 during my day job and had initially tried running this project that way. When it didn't work I set up my Windows dev env and copied the entire project to my windows root. I thought I'd cleaned up all the extraneous file permissions issues. I've fixed that change as well.

File: src/NzbDrone.Core/ImportLists/Plex/PlexParser.cs
Comment (vanstinator) [reply src/NzbDrone.Core/ImportLists/Plex/PlexParser.cs:45]: I've updated the parser with the new guids we get from the watchlist endpoint.

**Comment by vanstinator** [comment]
@Qstick how can I help this PR get merged?

**Comment by Qstick** [comment]
> @Qstick how can I help this PR get merged?

I'll see about running through it tonight

File: src/NzbDrone.Core/ImportLists/ImportListType.cs
Comment (Qstick) [inline src/NzbDrone.Core/ImportLists/ImportListType.cs:10]: Given these are just used for ordering, We should likely move Plex up above Advanced.

**Review by Qstick — CHANGES_REQUESTED** [review]
@vanstinator Code looks pretty good. I'll give it a test run tomorrow and merge from there. One nit on the ordering of Plex in the list types.

**Comment by vanstinator** [comment]
Awesome. I'll make that change tomorrow first thing!

File: src/NzbDrone.Core/ImportLists/ImportListType.cs
Comment (vanstinator) [reply src/NzbDrone.Core/ImportLists/ImportListType.cs:10]: Looking at this now I wonder if moving it above `Other` makes sense. Keep the first-party lists together, the random lists together, and then the advanced lists at the bottom

**Comment by Qstick** [comment]
@vanstinator did you ever push that one change above?

**Comment by vanstinator** [comment]
@Qstick I did. But apparently I forgot to commit and push it. https://github.com/Radarr/Radarr/pull/5707/commits/ff4ee3bf7c4fa68d3aec7e3c6cf131eab79db116

**Comment by vanstinator** [comment]
This should be good to go now @Qstick

**Comment by bakerboy448** [comment]
looks like you may need a rebase?
image

(Don't click update branch it merges the main branch...very messy, just rebase :) )

**Comment by austinwbest** [comment]
> looks like you may need a rebase?
> image
> 
> (Don't click update branch it merges the main branch...very messy, just rebase :) )

That is from the requested change not being marked as resolved..

I am curious if you have tried this outside of the plex agent? Other agents do not use the Guid array but instead a guid string.

**Comment by vanstinator** [comment]
> That is from the requested change not being marked as resolved..

It's because the review was marked as `Changes Requested` and the reviewer needs to update their review now that I've made the changes. I _could_ dismiss the review. But I generally prefer not to do that as the author.

> I am curious if you have tried this outside of the plex agent? Other agents do not use the Guid array but instead a guid string.

Are you talking about this line? https://github.com/Radarr/Radarr/pull/5707/files#diff-78ba780fd3580007ffcccc19ae48c988c3ef1ffaa77fd3e92a7f7f789f098232R38

This is a deserialized plex api response. This `Guids` array is completely unrelated to the `guid` string elsewhere in the code. I'm extracting a known guid type from the api response so that Radarr knows what to do with it.

**Comment by austinwbest** [comment]
> > That is from the requested change not being marked as resolved..
> 
> It's because the review was marked as `Changes Requested` and the reviewer needs to update their review now that I've made the changes. ~I _could_ dismiss the review.~ (I can't. That requires admin rights) But I generally prefer not to do that as the author.
> 
> > I am curious if you have tried this outside of the plex agent? Other agents do not use the Guid array but instead a guid string.
> 
> Are you talking about this line? https://github.com/Radarr/Radarr/pull/5707/files#diff-78ba780fd3580007ffcccc19ae48c988c3ef1ffaa77fd3e92a7f7f789f098232R38
> 
> This is a deserialized plex api response. This `Guids` array is completely unrelated to the `guid` string elsewhere in the code. I'm extracting a known guid type from the api response so that Radarr knows what to do with it.

I'm aware what needs to be done with the changes, I was letting baker know.

I'm also aware what you are doing with it, but my question still remains. If the agent is not plex, is the payload the same? The xml for non plex agents is not the same for other library types, so the question was related to testing the watch list type outside of the plex agent to make sure it is also not different because if it is, that would need to be looked at is all bud.

**Comment by vanstinator** [comment]
Oh I see what you're asking. The watchlist is coming from the cloud. So the library agent shouldn't factor in here at all which is nice as like you said that can be a minefield.

**Comment by jsclayton** [comment]
Any update on this? Would love to integrate this into my ingestion workflow. ❤️

**Comment by vanstinator** [comment]
Wonder why there's a sudden interest 😛

**Comment by igrschmidt** [comment]
> Wonder why there's a sudden interest 😛

Plex Discover it's going to make this feature SO valuable!

**Comment by PearsonFlyer** [comment]
Please add this Sonarr commit, and mark Q's open change request as "resolved".

https://github.com/Sonarr/Sonarr/commit/6e271e9272d6a109bc4b8f3debe784d767ee7a63

**Comment by bakerboy448** [comment]
> Please add this Sonarr commit, and mark Q's open change request as "resolved".
> 
> 
> 
> https://github.com/Sonarr/Sonarr/commit/6e271e9272d6a109bc4b8f3debe784d767ee7a63
> 
> 

That change should be cherry-picked to keep the original author and then marked as closing Radarr GHI #7188 (Bot Pull Request for it) and #7187 (Bot GitHub Issue for the  Commit)

**Comment by beyondmeat** [comment]
@vanstinator can you get the cherrypick in?

or could someone who can like @Qstick can you quickly do that to merge this in?

Sitting for 2 days and we are getting impatient lol.

**Comment by vanstinator** [comment]
> Sitting for 2 days and ppl are getting anxious and impatient lol.

I opened this PR _over a year ago_ 😉 

@PearsonFlyer done @bakerboy448 done

**Comment by Qstick** [comment]
Pushed a few cleanups to keep us in line with Sonarr on some things, plus the `[TestFixture]` Attribute. 

Thanks for the PR

**Comment by centogram** [comment]
Thanks for the pushing this.  The Sonarr variant works great.

**Comment by andyincountry** [comment]
I agree, I can't wait for this to come to the stable channel and be pushed out!!  Thank you very much for working on this!!

**Comment by AndyHazz** [comment]
Thinking ahead ... Sonarr only runs the Import List Sync task once per day which is a bit slow if you want to see things show up sooner. I expect Radarr will use the List Update Interval setting which has a minimum of 6 hours - better but still seems a bit limited?

As a workaround, using cron you can trigger a list sync from the api as often as you like - maybe there's a downside I'm not aware of? Perhaps as your watchlist in plex grows it would need longer between runs. I've been using Ofelia in my docker setup to do this without any obvious issues for a while. 

```
  ofelia:
    container_name: ofelia
    image: mcuadros/ofelia:latest
    depends_on:
      - sonarr
    command: daemon --docker
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
    restart: unless-stopped
```

... then add some labels to the Sonarr/Radarr container, e.g:

```
    labels:
      ofelia.enabled: "true"
      ofelia.job-exec.sonarr-import-watchlist.schedule: "@every 15m"
      ofelia.job-exec.sonarr-import-watchlist.command: "curl -d '{name: \"ImportListSync\"}' -H 'Content-Type: application/json' -X POST http://[sonarr ip address]:8989/api/v3/command/?apikey=[api key]"
```

It's a bit of a sledgehammer approach though - it would be nice if there was a better way to control this.

**Comment by ta264** [comment]
We have no plans to reduce the interval lower than 6 hours.  If every Radarr user hit the Plex API every 15 minutes I don't think it would be very long before Radarr got blocked.  So be a good citizen, and don't what @AndyHazz is recommending above.

**Comment by vanstinator** [comment]
I can confirm that we will take measures to more aggressively rate limit that endpoint if it gets abused.

**Comment by widgeen** [comment]
@vanstinator - I never even considered that this would be hitting a public endpoint. From a technical perspective, I get it, but would there be any possible option to look at moving wishlists over to something held locally rather than centerally?

If not, have you considered allowing subscriptions to a push notification for anything added to the watchlist rather than making this an API query?

**Comment by madAndroid** [comment]
Really keen to start using this - but it seems like the latest image doesnt' include the functionaility ... Is there any way to track when this will land in the stable docker image?

**Comment by bakerboy448** [comment]
This is only in nightly.

Watch the releases on GitHub or Reddit for announcements of the next release and it'd be included in the change log

**Comment by contact411** [comment]
The wait for this feature feels eternal!

**Comment by JStrongy** [comment]
Wait has been so long I don’t even think it’s needed anymore.

Plex have since added in the ability to obtain an RSS feed for the Watchlist, which can be added into Radarr as-is.

**Comment by AndyHazz** [comment]
Thanks for pointing that out! RSS works just fine, I hadn't noticed that feature before but it's documented here for reference: https://support.plex.tv/articles/universal-watchlist/

**Comment by RobinDadswell** [comment]
For information this is already in nightly

**Comment by loctruong96** [comment]
> Thanks for pointing that out! RSS works just fine, I hadn't noticed that feature before but it's documented here for reference: https://support.plex.tv/articles/universal-watchlist/

tried this and it works perfectly. thank you.

**Pull request merged**
